bundles/framework/layerlist/view/LayerList/LayerCollapse/CollapseService.js

```diff
@@ -0,0 +1,206 @@
+import { StateHandler, mutatorMixin } from 'oskari-ui/util';
+import { groupLayers } from './util';
+
+export const GROUPING_METHODS = {
+    ORGANIZATION: 'getOrganizationName',
+    GROUP: 'getInspireName'
+};
+
+const MIN_SEARCH_TEXT_LENGTH = 2;
+const ANIMATION_TIMEOUT = 400;
+const LAYER_REFRESH_THROTTLE = 2000;
+
+/**
+ * Holds and mutates layer list state.
+ * Handles events related to layer listing.
+ */
+class UIService extends StateHandler {
+    constructor (instance) {
+        super();
+        this.sandbox = instance.getSandbox();
+        this.mapLayerService = this.sandbox.getService('Oskari.mapframework.service.MapLayerService');
+        this.map = this.sandbox.getMap();
+        this.groupingMethod = null;
+        this.filter = {
+            activeId: null,
+            text: null
+        };
+        this.state = {
+            groups: [],
+            openGroupTitles: [],
+            selectedLayerIds: this._getSelectedLayerIds(),
+            mapSrs: this.map.getSrsName()
+        };
+        this.eventHandlers = this._createEventHandlers();
+    }
+    setGroupingMethod (groupingMethod) {
+        if (this.groupingMethod === groupingMethod) {
+            return;
+        }
+        this.groupingMethod = groupingMethod;
+        this.updateLayerGroups();
+    }
+    setFilter (activeId, searchText) {
+        this.filter = { activeId, searchText };
+        this.updateLayerGroups();
+    }
+    _getSelectedLayerIds () {
+        return this.map.getLayers().map(layer => layer.getId());
+    }
+
+    addLayer (id) {
+        if (!id || this.state.selectedLayerIds.includes(id)) {
+            return;
+        }
+        const selectedLayerIds = [...this.state.selectedLayerIds, id];
+        this.updateState({ selectedLayerIds });
+        setTimeout(() => this.sandbox.postRequestByName('AddMapLayerRequest', [id]), ANIMATION_TIMEOUT);
```

Comment:
Timeout is used for performance reasons. If UI is updated for selected layer during animation there is a notable performance hit and the animation becomes stuttery. This should be commented on the code as my first reaction would be to remove the timeout :)

Reply:
Added comment